numpy/_core/_add_newdocs.py

```diff
@@ -912,7 +912,7 @@
 
 add_newdoc('numpy._core.multiarray', 'asarray',
     """
-    asarray(a, dtype=None, order=None, *, like=None)
+    asarray(a, dtype=None, order=None, *, device=None, like=None)
```

Comment:
Also, why not just set the default to `"cpu"`? That seems clearer. (Still fine to accept `None`).

Reply:
Right now the function dispatch requires `like` to be the last argument - here's a check that raises `RuntimeError` if a function doesn't comply with it: https://github.com/numpy/numpy/blob/448e4dabb5e7d593af7d164a335dc7e44e17ec95/numpy/_core/overrides.py#L149-L156

Can I keep `device` before `like` or should we update the `__array_function__` protocol?

Reply:
> Also, why not just set the default to `"cpu"`? That seems clearer. (Still fine to accept `None`).

That's true that `"cpu"` is the only supported value, but I would prefer to stick to the Array API standard here. It defines the default value for `device` as `None`: https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html

We're adding these keyword arguments for Array API compatibility only, so I would prefer to fully conform with it.

Reply:
I don't think it matters from the Array API perspective, it doesn't make NumPy non-compliant, it just makes it explicit about what that default is.

But, I'll give a very slight vote to just keep it `None` anyway.  Just becuase I don't think the more explicit `device="cpu"` makes it clearer, since users should ignore it anyway (you only use it via `xp = arr.__array_namespace__()` as an Array API user, never directly as a NumPy user).

Reply:
The order doesn't matter much in principle, since it's keyword-only and `like` is only for other non-numpy libraries as well. However, there's a bigger problem here:
```python
>>> import numpy as np
>>> import dask.array as da
>>> d = da.ones((1, 2))
>>> np.asarray([1, 2], device='cpu')
array([1, 2])
>>> np.asarray([1, 2], device='cpu', like=d)
Traceback (most recent call last):
  Cell In[24], line 1
    np.asarray([1, 2], device='cpu', like=d)
  File ~/code/tmp/dask/dask/array/core.py:1760 in __array_function__
    return da_func(*args, **kwargs)
  File ~/code/tmp/dask/dask/array/core.py:4581 in asarray
    return from_array(a, getitem=getter_inline, **kwargs)
TypeError: from_array() got an unexpected keyword argument 'device'
```

The `device` keyword probably should not be added to the dispatchers for any of these creation functions. It looks like the `__array_function__` protocol wasn't implemented in a forward-looking way - I haven't looked much deeper yet, but at first sight it may be the case that we can never add a new keyword to dispatched functions anymore.

In this case, `device` is here to support writing cross-library-compatible code via the array API standard. The dispatching protocols kinda were earlier attempts to do something similar. So probably they should not interact anyway?

Reply:
@rgommers - I've gotten quite used to updating astropy's `__array_function__` wrappers to remain compatible with numpy functions -- and we've got tests to check that our signatures remain consistent. Of course it doesn't happen that often. So, I don't think that's a worry, and overall I'd tend to not special-case any arguments (i.e., pass it on if present).

On the order: if `like` has to be last, then just keep it like that, sorry for not noticing myself.

Reply:
Just to clarify: the check is only there to test habits, and probably mainly to make sure it is kwarg only.
It may have to be last in C (not sure), but it doesn't really matter anywhere here.  But... it also doesn't matter a lot (especially since if someone cares, they can change it).

Reply:
Thanks @mhvk, that is useful context. A test that flag mismatches sounds like a great idea. 🤞🏼 most implementers have that.